### PR TITLE
gh-94673: Add Per-Interpreter tp_weaklist for Static Builtin Types

### PR DESCRIPTION
**Pull request by ericsnowcurrently**
* Issue: gh-94673

### CODE REVIEW

File: Objects/weakrefobject.c [hunk 1]
```diff
@@ -1018,3 +1018,19 @@ PyObject_ClearWeakRefs(PyObject *object)
...
+ * This is called at the end of runtime finalization, so we can just
+ * wipe out the type's weaklist.  We don't bother with callbacks
+ * or anything else.
+ */
+void
+_PyStaticType_ClearWeakRefs(PyTypeObject *type)
+{
+    static_builtin_state *state = _PyStaticType_GetState(type);
+    PyObject **list = _PyStaticType_GET_WEAKREFS_LISTPTR(state);
+    while (*list != NULL) {
```
Comment (kumaraditya303) [inline Objects/weakrefobject.c:1033]: This should be `if` rather than `while`, no?

Comment (ericsnowcurrently) [reply Objects/weakrefobject.c:1033]: I agree it is confusing.  However, it is how `clear_weakref()` works.  `*list` is a pointer to the head of the linked list and `list` is a pointer to that pointer.  `clear_weakref()` actually modifies `list`, changing where `*list` points.  Thus each time through the loop we are actually popping off the next node and clearing it.

A short comment here wouldn't hurt, even if `clear_weakref()` is fundamentally confusing everywhere.  (I'll open a separate issue about that.)

File: Doc/whatsnew/3.12.rst [hunk 2]
```diff
@@ -413,6 +413,13 @@ Porting to Python 3.12
   ``Py_UNICODE*`` based format (e.g. ``u``, ``Z``) anymore. Please migrate
   to other formats for Unicode like ``s``, ``z``, ``es``, and ``U``.
 
+* ``tp_weaklist`` for ``PyType_Type`` is always ``NULL`` for all static
+  builtin types.  This is an internal-only field on ``PyTypeObject``
```
Comment (ericsnowcurrently) [inline Doc/whatsnew/3.12.rst]:
```suggestion
* ``tp_weaklist`` for all static builtin types is always ``NULL``.
   This is an internal-only field on ``PyTypeObject``
```

**Comment by markshannon** [comment]
@ericsnowcurrently 
Have you benchmarked this?

I expect that it won't make any difference, but I'm a bit concerned as `_PyObject_GET_WEAKREFS_LISTPTR` is on the "fast" path of normal object de-allocation. I put "fast" in quotes because it is far from fast, but I don't want to make it any slower.

It seems to me, that the extra check is not needed during de-allocation.

**Comment by ericsnowcurrently** [comment]
Good point.  I'll take a look Monday.

**Comment by ericsnowcurrently** [comment]
Pyperformance shows 1.01x slower, so I'll need to address that.

**Pull request merged**